### PR TITLE
Remove deprecated `spec.eni.{min-allocate,pre-allocate,max-above-watermark}` parameters

### LINKED ISSUE #21858: Remove deprecated `spec.eni.{min-allocate,pre-allocate,max-above-watermark}` parameters
When we added Azure IPAM support in Cilum 1.7, we moved certain IPAM knobs (e.g. `spec.eni.pre-allocate/`) from `spec.eni` to `spec.ipam`. https://github.com/cilium/cilium/pull/10089/commits/cd24805873ad1baa1c23d26d96c8dbb4262be629

The following knobs have been deprecated since Cilium v1.7 and should be removed:

```
spec.eni.min-allocate
spec.eni.pre-allocate
spec.eni.max-above-watermark
```

### PR DESCRIPTION
**Pull request by oxxenix**
Dropped deprecated spec.eni parameters in the codebase.

Fixes: #21858

### CODE REVIEW

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]: This does not build, `nil` cannot be returned in a function returning an `int`. I think you can just return the value here, no need for the if anymore statement.

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]: Please remove this whitespace

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]: Please remove this whitespace.

File: pkg/ipam/node.go
Comment (gandro) [inline pkg/ipam/node.go]: Same as above, `nil` cannot be returned here. The if statement is also not necessary anymore.

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks for the PR. There are a few issues that need to be addressed, please take a look at my comments.

File: pkg/ipam/node.go
Comment (oxxenix) [reply pkg/ipam/node.go]: Thank you for your comments. Fixed now :)

**Review by gandro — COMMENTED** [review]
Thanks! Checking all usages in my IDE, there are some more usages here that can be removed:

https://github.com/cilium/cilium/blob/927bd8c26904ff92e42c61cec6d00ea8ac062c05/pkg/nodediscovery/nodediscovery.go#L573

Both for MinAllocate and PreAllocate.


Let's also update this tests to not use the legacy "PreAllocate" field anymore:
https://github.com/cilium/cilium/blob/32d2ef8e0445a79b1979312f33e0b514ca7650a5/plugins/cilium-cni/types/types_test.go#L128

File: plugins/cilium-cni/types/types_test.go
Comment (gandro) [inline plugins/cilium-cni/types/types_test.go]: This is not deprecated. Please keep this in.

File: plugins/cilium-cni/types/types_test.go
Comment (gandro) [inline plugins/cilium-cni/types/types_test.go]: It seems there is an indentation change here, this line should not be modified.

File: pkg/nodediscovery/nodediscovery.go
Comment (gandro) [inline pkg/nodediscovery/nodediscovery.go:582]: Please do not remove this newline.

File: plugins/cilium-cni/types/types_test.go
Comment (gandro) [inline plugins/cilium-cni/types/types_test.go]: Same as above, this is not deprecated. Please keep this in.

File: pkg/nodediscovery/nodediscovery.go
Comment (gandro) [inline pkg/nodediscovery/nodediscovery.go]: There is now some trailing tabs after the re-introduced newline, please remove the tabs, but keep the newline.

File: plugins/cilium-cni/types/types_test.go
Comment (oxxenix) [reply plugins/cilium-cni/types/types_test.go]: this one looks okay in my IDE. Not sure why here looks like the indentation problem

File: plugins/cilium-cni/types/types_test.go
Comment (gandro) [inline plugins/cilium-cni/types/types_test.go]: There are white-space changes here, please undo them.

File: plugins/cilium-cni/types/types_test.go
Comment (gandro) [inline plugins/cilium-cni/types/types_test.go]: Same as above, there are white-space changes here, please undo them.

Comment (oxxenix) [reply plugins/cilium-cni/types/types_test.go]: Yes, I noticed them as well after submitting the commit. Looks ok in my IDE though. Will check once again

Comment (oxxenix) [reply plugins/cilium-cni/types/types_test.go]: I am working on it. Will ping when its done

**Comment by nbusseneau** [comment]
Converting to draft, do not hesitate to mark as ready for review when happy with the changes :)

**Comment by gandro** [comment]
Not sure what's happening with maintainers-little-helper, the commits seem signed. Feel free to ignore.

File: plugins/cilium-cni/types/types_test.go
Comment (oxxenix) [reply plugins/cilium-cni/types/types_test.go]: @gandro, @nbusseneau fixed. Please, take a look.

**Review by gandro — COMMENTED** [review]
Thanks! This is looking solid now.

I forgot this earlier: Do you mind adding a small note to the `upgrade.rst` docs mentioning that these deprecated fields have been removed?

For example:
```diff
--- a/Documentation/operations/upgrade.rst
+++ b/Documentation/operations/upgrade.rst
@@ -312,6 +312,11 @@ Annotations:
 1.13 Upgrade Notes
 ------------------
 
+* The code for the deprecated ``spec.eni.min-allocate``, ``spec.eni.pre-allocate``
+  ``spec.eni.max-above-watermark`` fields has been removed, those fields are
+  no longer functional. Please use their semantically equivalent ``spec.ipam``
+  replacements instead.
+
 Removed Options
 ~~~~~~~~~~~~~~~
 
```

There is an open question if we should also remove the CiliumNode CRD fields completely, but that might cause upgrade issues. I defer to @cilium/sig-k8s on that.

**Comment by obaranov1** [comment]
@gandro Is the note should be added at the end of the upgrade.rst document?

**Comment by gandro** [comment]
> @gandro Is the note should be added at the end of the upgrade.rst document?

Just where it is in my patch, i.e. directly under `1.13 Upgrade Notes` as a bullet point.

**Comment by qmonnet** [comment]
The changes look good, thank you!

I've got some comments on the commit format:

image

Can you please use something more descriptive for the object? Maybe `eni: Remove deprecated spec.eni parameters`. This is so we can understand better what the commit does when looking at the Git log. If you want to link to the GH issue, you can instead do it in the commit description. Last thing, the Signed-off-by tag should normally contain your full name rather than GH handle. So ideally, your commit title and description could look like this (assuming your email contains your name):

```
eni: Remove deprecated spec.eni parameters

These parameters have been deprecated since v1.9, let's clean them up.

Fixes: https://github.com/cilium/cilium/issues/21858
Signed-off-by: Oksana Baranova <oksana.baranova@intel.com>
```

File: Documentation/operations/upgrade.rst
Comment (qmonnet) [inline Documentation/operations/upgrade.rst]: I missed the `*` at the beginning, but the CI caught it. Please indent the body of the paragraph to align it with the beginning of the text to make it a list item, the build for the docs does not pass otherwise.

```suggestion

* The code for the deprecated ``spec.eni.min-allocate``, ``spec.eni.pre-allocate``
  ``spec.eni.max-above-watermark`` fields has been removed, those fields are
  no longer functional. Please use their semantically equivalent ``spec.ipam``
  replacements instead.
```

The Go checks and Travis jobs complain on the trailing tab characters raised by Sebastian, please address those as well.

**Review by qmonnet — CHANGES_REQUESTED** [review]

**Comment by obaranov1** [comment]
When you execute `git commit -s -m "message"`, Git automatically creates a Signed-off by taking Github username and email. But I updated the signed-off according to requirements.

**Comment by qmonnet** [comment]
As far as I know, `git commit -s` creates the tag with the user name set for the repository (configurable with `git config user.name 'Quentin Monnet'` for example). I suppose it's somehow configured with your handle in your repo. But OK, thanks a lot!

Looks like the trailing tabs are still here though. Also, any chance you could squash the two commits (it seems there's a new one that appeared when you accepted my suggestion?) together please?

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR. There's some go linting errors that need fixing.

**Comment by obaranov1** [comment]
@christarazi, Hello! Regarding the go linting errors, how they can be fixed?

File: pkg/ipam/crd.go
Comment (qmonnet) [inline pkg/ipam/crd.go]: This line has a trailing tab character at the end.

File: pkg/ipam/node.go
Comment (qmonnet) [inline pkg/ipam/node.go]: This line as well.

**Review by qmonnet — COMMENTED** [review]
The Go linting errors are the trailing tab characters mentioned above (and repeated below for convenience). The lint workflow runs `go fmt` on the source and complains if anything differs.

File: pkg/ipam/crd.go
Comment (oxxenix) [reply pkg/ipam/crd.go]: thank you! will fix them

File: pkg/ipam/crd.go
Comment (qmonnet) [inline pkg/ipam/crd.go]: This one still has a tab at the end. I probably missed it on my earlier comment, sorry.

**Comment by tommyp1ckles** [comment]
@obaranov1 Thanks for the PR 😃 

Just searching through I see some potential other places these might need to be cleaned up:

* [`examples/crds/v2/ciliumnodes.yaml`](https://github.com/cilium/cilium/blob/master/pkg/k8s/apis/cilium.io/client/crds/v2/ciliumnodes.yaml#L188)
* https://github.com/cilium/cilium/blob/fb81f90e91e75130e4c75838552240cf87cc8ad4/Documentation/network/concepts/ipam/eni.rst

**Comment by jrajahalme** [comment]
/test-1.25-net-next

**Review by gandro — APPROVED** [review]

**Review by qmonnet — APPROVED** [review]
Thanks!

**Review by tommyp1ckles — APPROVED** [review]

**Review by christarazi — APPROVED** [review]

**Comment by christarazi** [comment]
/test-1.26-net-next

**Comment by gandro** [comment]
net-next failed on a known-flaky test https://github.com/cilium/cilium/pull/22403

**Review by gandro — APPROVED** [review]

**Pull request merged**
